### ISSUE TITLE
`uuid.UUID` or `str` in SQLAlchemy models

### ISSUE THREAD

**Issue by rdimaio** [rucio/rucio#6743]
This is a follow-up from this PR comment: https://github.com/rucio/rucio/pull/6497#discussion_r1494426294, where it was suggested to change all instances of `Mapped[uuid.UUID]` to `Mapped[str]`. 

## `uuid.UUID`
This is the Python-implemented `UUID` type: https://docs.python.org/3/library/uuid.html#uuid.UUID, which is the type hint we're currently using, e.g.:

https://github.com/rucio/rucio/blob/2c389be05cd37ab81cae35e4f49ff905a5efa00f/lib/rucio/db/sqla/models.py#L1443-L1444

## `str`

There is this SO answer where this is discussed https://stackoverflow.com/a/41434923, and there are a couple comments that mention some downsides to storing UUIDs as strings:

> there are many downsides - a few off the top of mind: size - string uuid takes up twice the space - 16bytes vs 32 chars - not including any formatters. Processing time - more bytes = more processing time by the CPU as your dataset gets bigger. uuid string formats differ by language, adding addition required translations. Easier for someone to misuse the column, as you can put anything in there, things that aren't uuids.

## `sqlalchemy.types.Uuid`

As an alternative to both `uuid.UUID` and `str`, there is the SQLA2.0 `Uuid` type: https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid

>  class sqlalchemy.types.Uuid
    Represent a database agnostic UUID datatype.
    For backends that have no “native” UUID datatype, the value will make use of CHAR(32) and store the UUID as a 32-character alphanumeric hex string.
    For backends which are known to support UUID directly or a similar uuid-storing datatype such as SQL Server’s UNIQUEIDENTIFIER, a “native” mode enabled by default allows these types will be used on those backends.

This seems to be a similar implementation to the `GUID` type we're currently using: https://github.com/rucio/rucio/blob/2c389be05cd37ab81cae35e4f49ff905a5efa00f/lib/rucio/db/sqla/types.py#L28-L51

See also this SO answer: https://stackoverflow.com/a/76059829

## To do
I think `sqlalchemy.types.Uuid` seems to be the appropriate type to use in this context. However, I think it might be worth investigating this a bit further before we commit to a specifc type.

**Comment by dchristidis** [comment]
> This is the Python-implemented UUID type: https://docs.python.org/3/library/uuid.html#uuid.UUID, which is what we're currently using.

To avoid any misunderstanding, this is what the type hints claim we are using. As far as I can tell, we are not.

> There is this SO answer where this is discussed https://stackoverflow.com/a/41434923, and there are a couple comments that mention some downsides to storing UUIDs as strings:

I’m not sure if this particular comment is relevant because we do not store UUIDs as strings in the database.

> sqlalchemy.types.Uuid

At some point, we will have to convert this to a string (e.g. in the gateway or REST layer).

**Comment by rdimaio** [comment]
> > This is the Python-implemented UUID type: https://docs.python.org/3/library/uuid.html#uuid.UUID, which is what we're currently using.
> 
> To avoid any misunderstanding, this is what the type hints claim we are using. As far as I can tell, we are not.

Thanks for the correction on the first point, I updated the main post.

> > There is this SO answer where this is discussed https://stackoverflow.com/a/41434923, and there are a couple comments that mention some downsides to storing UUIDs as strings:
> 
> I’m not sure if this particular comment is relevant because we do not store UUIDs as strings in the database.
> 

Good point. I mistakenly conflated using `str` as the Mapped type hint in the SQLA models with the type that would actually end up being stored in the database.

> > sqlalchemy.types.Uuid
> 
> At some point, we will have to convert this to a string (e.g. in the gateway or REST layer).

The [`sqlalchemy.types.Uuid.__init__`](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid.__init__)`as_uuid` parameter (https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid.params.as_uuid) mentions: "if True, values will be interpreted as Python uuid objects, converting to/from string via the DBAPI.", which seems to indicate that it allows for conversion to string in a DB-agnostic way

**Comment by dchristidis** [comment]
> The [`sqlalchemy.types.Uuid.__init__`](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid.__init__)`as_uuid` parameter (https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid.params.as_uuid) mentions: "if True, values will be interpreted as Python uuid objects, converting to/from string via the DBAPI.", which seems to indicate that it allows for conversion to string in a DB-agnostic way

Do I understand correctly that you are suggesting replacing our own `GUID` class with this `Uuid`, in a way like:

```python
rse_id: Mapped[str] = mapped_column(Uuid(as_uuid=False))
```

**Comment by rdimaio** [comment]
> > The [`sqlalchemy.types.Uuid.__init__`](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid.__init__)`as_uuid` parameter (https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid.params.as_uuid) mentions: "if True, values will be interpreted as Python uuid objects, converting to/from string via the DBAPI.", which seems to indicate that it allows for conversion to string in a DB-agnostic way
> 
> Do I understand correctly that you are suggesting replacing our own `GUID` class with this `Uuid`, in a way like:
> 
> ```python
> rse_id: Mapped[str] = mapped_column(Uuid(as_uuid=False))
> ```

Yeah, I think it might be better in the long run to move to their version of `Uuid` instead of maintaining our own implementation of this similar DB-agnostic logic. Also, here's the SQLA source code to double_check that the type of `Uuid` is `str` when `as_uuid=False`:

https://github.com/sqlalchemy/sqlalchemy/blob/f53f261ad85fb251ac8989d7d9b0ccf1c099a892/lib/sqlalchemy/sql/sqltypes.py#L3532-L3544

and the`python_type` property:

https://github.com/sqlalchemy/sqlalchemy/blob/f53f261ad85fb251ac8989d7d9b0ccf1c099a892/lib/sqlalchemy/sql/sqltypes.py#L3565-L3567

**Comment by hyperknot** [comment]
Hi, based on this discussion and this cookbook, I'm trying to come up with a recommended UUIDv7 PK type.


```
import uuid
from typing import Annotated, TypeAlias

import uuid_utils
from sqlalchemy import Uuid
from sqlalchemy.orm import mapped_column


PK_UUID: TypeAlias = Annotated[
    uuid.UUID, mapped_column(Uuid, primary_key=True, default=uuid_utils.uuid7)
]
```

I mean the left side is stdlib's UUID, the right side is sqlalchemy.Uuid (not sqlalchemy.types.uuid, or is this an alias?). For generation I'm using uuid_utils.uuid7, until Python 3.14 gets more supported in the ecosystem.

Can you tell me if this is correct?

**Comment by dchristidis** [comment]
If I am not mistaken, the `Uuid` class appears to be implemented as a string on Oracle. If so, then it is not an attractive option.

```python
>>> from sqlalchemy.dialects.oracle.base import OracleDialect
>>> OracleDialect.supports_native_uuid
False
```

https://github.com/sqlalchemy/sqlalchemy/blob/adfec9e30e9178d7af1de5417b5edf0aee136934/lib/sqlalchemy/sql/sqltypes.py#L3754-L3779

**Comment by bari12** [comment]
Thank you for raising this. We have reviewed this issue as part of our regular issue review.

We see value in this request, but we do not have the capacity to work on it within the next twelve months.

In line with our [issue management guideline](https://rucio.cern/documentation/developer/issue_management), we are labelling it `wishlist` and closing it. This does not mean it is forgotten: wishlist items are revisited regularly during component reviews and roadmap planning, and this issue can be re-opened and added back to the pipeline as soon as interest and capacity align.
